Sources/SnapshotTesting/Snapshotting/UIImage.swift

```diff
@@ -189,6 +171,51 @@ private func diff(_ old: UIImage, _ new: UIImage) -> UIImage {
 import CoreImage.CIKernel
 import MetalPerformanceShaders
 
+@available(iOS 10.0, tvOS 10.0, macOS 10.13, *)
+func perceptuallyCompare(_ old: CIImage, _ new: CIImage, pixelPrecision: Float, perceptualPrecision: Float) -> String? {
+  let deltaOutputImage = old.applyingFilter("CILabDeltaE", parameters: ["inputImage2": new])
+  let thresholdOutputImage: CIImage
+  do {
+    thresholdOutputImage = try ThresholdImageProcessorKernel.apply(
+      withExtent: new.extent,
+      inputs: [deltaOutputImage],
+      arguments: [ThresholdImageProcessorKernel.inputThresholdKey: (1 - perceptualPrecision) * 100]
+    )
+  } catch {
+    return "Newly-taken snapshot's data could not be loaded. \(error)"
+  }
+  var averagePixel: Float = 0
+  let context = CIContext(options: [.workingColorSpace: NSNull(), .outputColorSpace: NSNull()])
+  context.render(
+    thresholdOutputImage.applyingFilter("CIAreaAverage", parameters: [kCIInputExtentKey: new.extent]),
+    toBitmap: &averagePixel,
+    rowBytes: MemoryLayout<Float>.size,
+    bounds: CGRect(x: 0, y: 0, width: 1, height: 1),
+    format: .Rf,
+    colorSpace: nil
+  )
+  let actualPixelPrecision = 1 - averagePixel
+  guard actualPixelPrecision < pixelPrecision else { return nil }
+  var maximumDeltaE: Float = 0
+  context.render(
+    deltaOutputImage.applyingFilter("CIAreaMaximum", parameters: [kCIInputExtentKey: new.extent]),
+    toBitmap: &maximumDeltaE,
+    rowBytes: MemoryLayout<Float>.size,
+    bounds: CGRect(x: 0, y: 0, width: 1, height: 1),
+    format: .Rf,
+    colorSpace: nil
+  )
+  let actualPerceptualPrecision = 1 - maximumDeltaE / 100
```

Comment:
@ejensen I'm not familiar with this filter's output, but when taking this for a spin in isowords I noticed that it seems to be flipped:

<img width="707" alt="image" src="https://user-images.githubusercontent.com/658/191859338-2340617f-d461-4ba6-91b7-06589983ec8b.png">

Should this just be `maximumDeltaE / 100`?

Reply:
Or maybe not, I was misremembering for this one:

<img width="676" alt="IMG_3696" src="https://user-images.githubusercontent.com/658/191859557-aab5819f-3daa-4ff5-a0b4-20b0c976490c.png">

Might just need to get a feel for the change and what the numbers "mean" :)

Reply:
@stephencelis Delta E should have an inverse relationship with precision. A higher Delta(Difference)E means the pixels are less similar(precise).

The wording of the error reporting might need some finessing. Seeing  "Actual perceptual precision: 0.56" insinuates the entire image is significantly different. When in reality the perceptual precision being reported is the lowest of all the pixels compared, could be a significant outlier.

Reply:
So when it says "Actual image precision 0.03083378 is less than required 0.9" what does that mean?

Reply:
> So when it says "Actual image precision 0.03083378 is less than required 0.9" in my test logs what does that mean?

That means that the lowest perceptual precision of all the pixels is 3.08%. You would need to set `perceptualPrecision <= 0.0308` if you were to require all the pixels to meet the perceptual requirement  (eg. setting `precision: 1.0`).

The failure messages could both better explain the values and provide suggestions. I'm prototyping this in the [image-comparison-playground branch](https://github.com/pointfreeco/swift-snapshot-testing/compare/main...Frameio:swift-snapshot-testing:image-comparison-playground). It contains an iOS/macOS compatible Swift Playground where you supply two snapshots and it displays the result of the comparison along with all the intermediaries for inspection.

<img width="1449" alt="NotMatching" src="https://user-images.githubusercontent.com/915431/192423006-c30a8612-af2b-4b0f-9af2-735063727c79.png"><img width="1449" alt="Matching" src="https://user-images.githubusercontent.com/915431/192423003-b4beb410-9883-473a-bc54-d28c32e7463a.png">

The output makes suggestions for values to use when images do not match:
```
❌ Images do not match…
The images would match if either:
  1. precision is lowered to 0.9991
  2. perceptualPrecision is lowered to 0.9939
  3. precision is raised to 1.0 and perceptualPrecision is lowered to 0.9912
ℹ️ 99.91% of pixels match with 99.98% perceptual precision
```

Matching images also get suggestions when the precision values could be raised:
```
✅ Images match
The images would still match if:
  1. precision is raised to 0.9998
  2. perceptualPrecision is raised to 0.9939
  3. precision is raised to 1.0 and perceptualPrecision is set to 0.9859
ℹ️ 100.00% of pixels match with 98.00% perceptual precision
```

There is a tradeoff between detail + precision and speed of execution. It is particularly expensive to calculate the maximum matching `perceptualPrecision` when `precision < 1`. It takes a couple of seconds to calculate with a full-screen snapshot. The other suggestions are quite fast and are likely reasonable to include in the swift-snapshot-test's failure messages.

Reply:
> Actual image precision 0.03083378 is less than required 0.9

Isn't that referring to `precision`, not `perceptualPrecision`?

Reply:
So there is at least one pixel in the image that is 97% different from the same pixel on the reference imag?

Reply:
> So does it mean there is at least one pixel in the image that is 97% different from the same pixel on the reference image

Yes, that is what it means.

Reply:
I guess currently there's nothing built-in that I can use to filter out outliers?